### PR TITLE
Fix two opaque return type bugs

### LINKED ISSUE #56367: [SR-13972] Crash in emitDirectTypeMetadataRef -- trivial repro case
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-13972      |
   |Radar             | rdar://problem/72407668         |
   |Original Reporter | davidhun (JIRA User)      |
   |Type              | Bug    |

Attachment: [Download](https://user-images.githubusercontent.com/2727770/164964038-fce9d3e9-55b2-435f-83d3-f820dc2fc3e1.gz)

      
<details>
  <summary>Environment</summary>

MacOS: 11.1 (20C69)

Apple Swift version 5.3.2 (swiftlang-1200.0.45 clang-1200.0.32.28)

Target: x86_64-apple-darwin20.2.0

</details>

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 1         |
   |Component/s       | Compiler    |
   |Labels            | Bug, Crash        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: 2693568d8a0ddefb32d72affc752b757

  </details>

**Issue Description:**

The attached project with 2 source files consistently crashes the swift compiler.

To repeat:

1.  cd SILCrash

2.  swift build

### LINKED ISSUE #60038: Bugs with opaque result types in structural position
Each of these three declarations should type check, but doesn't:

```swift
protocol P {
}

extension P {
  func f1() -> () -> (some P) {
    return { self } // error: cannot convert value of type 'Self' to closure result type 'some P'
  }

  func f2() -> () -> (some P) {
    return { () -> Self in self } // error: function declares an opaque return type, but has no return statements in its body from which to infer an underlying type
  }

  func f3() -> (some P).Type {
    return Self.self // error: function declares an opaque return type, but has no return statements in its body from which to infer an underlying type
  }
}
```

### LINKED ISSUE #64799: Compiler crashes when face with particular usage of the `some` keyword in a closure type
**Description**

I just stumbled upon the very bad reaction that I received from the compiler/Xcode upon writing this function:

```swift
func foo () {
    var test: ()->some Equatable = { ()->Int in 0 }
}
```

If I paste this function into a fresh package I get only one error (not associated with any line of code): `Command CompileSwift failed with a nonzero exit code`, which shows me no additional detail underneath it when I click on it. Additionally, Xcode displays the message `An internal error occurred. Source editor functionality is limited. Attempting to restore...`.

**Steps to reproduce**

Attempt to compile the above function.

**Expected behavior**

"Build Succeeded" or a compilation error.

**Environment**
- Swift version 5.7.2 (swiftlang-5.7.2.135.5 clang-1400.0.29.51)

### PR DESCRIPTION
**Pull request by slavapestov**
- Fixes https://github.com/apple/swift/issues/56367 / rdar://72407668
- Fixes https://github.com/apple/swift/issues/60038 partially
- Fixes #64799

### CODE REVIEW

File: lib/Sema/CSApply.cpp [hunk 1]
```diff
@@ -6875,6 +6875,60 @@ Expr *ExprRewriter::coerceToType(Expr *expr, Type toType,
...
+    // Compute the substitutions for the opaque type declaration.
+    auto opaqueLocator = solution.getConstraintSystem().getOpenOpaqueLocator(
+        locator, opaqueDecl);
+    SubstitutionMap substitutions = solution.computeSubstitutions(
+        opaqueDecl->getOpaqueInterfaceGenericSignature(), opaqueLocator);
+
+    // If we don't have substitutions, this is an opaque archetype from
+    // another declaration being manipulated, and not an erasure of a
+    // concrete type to an opaque type inside its defining declaration.
+    if (!substitutions.empty()) {
```
Comment (slavapestov) [inline lib/Sema/CSApply.cpp:6908]: @hborla @xedin Is there a better way to check if this is the point of inference of the underlying type of an opaque return type, and not merely a usage of an opaque return type from a different context?

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: lib/Sema/CSApply.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CSApply.cpp:6908]: I don't think there is a better way to check this unfortunately...

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Pull request merged**